src/move/emojicoin_arena/sources/emojicoin_arena.move

```diff
@@ -0,0 +1,1037 @@
+        /// Number of melee-specific swaps.
+        n_swaps: Aggregator<u64>,
+        /// Volume of melee-specific swaps in octas.
+        swaps_volume: Aggregator<u128>,
+        /// Amount of emojicoin 0 locked in all melee escrows for the melee.
+        emojicoin_0_locked: Aggregator<u64>,
+        /// Amount of emojicoin 1 locked in all melee escrows for the melee.
+        emojicoin_1_locked: Aggregator<u64>
+    }
+
+    struct Registry has key {
+        /// A map of each melee's `melee_id` to the melee.
+        melees_by_id: SmartTable<u64, Melee>,
+        /// Map from a sorted combination of market IDs (lower ID first) to the melee serial ID.
+        melee_ids_by_market_ids: SmartTable<vector<u64>, u64>,
+        /// Approves transfers from the vault.
+        signer_capability: SignerCapability,
+        /// `Melee.duration` for next melee.
+        next_melee_duration: u64,
+        /// `Melee.available_rewards` for next melee.
+        next_melee_available_rewards: u64,
+        /// `Melee.max_match_percentage` for next melee.
+        next_melee_max_match_percentage: u64,
+        /// `Melee.max_match_amount` for next melee.
+        next_melee_max_match_amount: u64,
+        /// All entrants who have entered a melee.
+        all_entrants: SmartTable<address, Nil>,
+        /// Number of melee-specific swaps.
+        n_swaps: Aggregator<u64>,
+        /// Volume of melee-specific swaps in octas.
+        swaps_volume: Aggregator<u128>,
+        /// Amount of octas matched. Decremented when a user taps out.
+        octas_matched: u64
+    }
+
+    struct Escrow<phantom Coin0, phantom LP0, phantom Coin1, phantom LP1> has key {
+        /// Corresponding `Melee.melee_id`.
+        melee_id: u64,
+        /// Emojicoin 0 holdings.
+        emojicoin_0: Coin<Coin0>,
+        /// Emojicoin 1 holdings.
+        emojicoin_1: Coin<Coin1>,
+        /// Number of swaps user has executed during the melee.
+        n_swaps: u64,
+        /// Volume of user's melee-specific swaps in octas.
+        swaps_volume: u128,
+        /// Cumulative amount of APT entered into the melee since the most recent deposit into an
+        /// empty escrow. Inclusive of total amount matched from locking in since most recent
+        /// deposit into an empty escrow. Reset to 0 upon exit.
+        octas_entered: u64,
+        /// Cumulative amount of APT matched since most recent deposit into an empty escrow, reset
+        /// to 0 upon exit. Must be paid back in full when tapping out.
+        octas_matched: u64
+    }
+
+    struct UserMelees has key {
+        /// Set of serial IDs of all melees the user has entered.
+        entered_melee_ids: SmartTable<u64, Nil>,
+        /// Set of serial IDs of all melees the user has exited.
+        exited_melee_ids: SmartTable<u64, Nil>,
+        /// Set of serial IDs of all melees the user has entered but not exited.
+        unexited_melee_ids: SmartTable<u64, Nil>
+    }
```

Comment:
@CRBl69 @xbtmatt after implementing all the onchain indexing logic and working with Tom on Figma, I tweaked a few fields - how does this finalized prototype look?

If we are good on these, and if Aptos Labs reviews, then we can probably start to scope out what the event fields look like